device/core/src/module_client.ts

```diff
@@ -182,26 +114,75 @@ export class ModuleClient extends EventEmitter {
    * @param callback Function to call when the operations have been queued.
    */
   sendOutputEventBatch(outputName: string, messages: Message[], callback: (err?: Error, result?: results.MessageEnqueued) => void): void {
-    this._internalClient.sendOutputEventBatch(outputName, messages, callback);
+    const retryOp = new RetryOperation(this._retryPolicy, this._maxOperationTimeout);
+    retryOp.retry((opCallback) => {
+      /* Codes_SRS_NODE_MODULE_CLIENT_18_011: [ The `sendOutputEventBatch` method shall send the list of events (indicated by the `messages` argument) via the transport associated with the Client instance. ]*/
+      this._transport.sendOutputEventBatch(outputName, messages, opCallback);
+    }, (err, result) => {
+      /*Codes_SRS_NODE_MODULE_CLIENT_18_021: [ When the `sendOutputEventBatch` method completes the `callback` function shall be invoked with the same arguments as the underlying transport method's callback. ]*/
+      /*Codes_SRS_NODE_MODULE_CLIENT_18_022: [ The `sendOutputEventBatch` method shall not throw if the `callback` is not passed. ]*/
+      safeCallback(callback, err, result);
+    });
+  }
+
+  close(closeCallback?: (err?: Error, result?: results.Disconnected) => void): void {
+    this._transport.removeListener('disconnect', this._moduleDisconnectHandler);
+    super.close(closeCallback);
+  }
+
+  private _disableInputMessages(callback: (err?: Error) => void): void {
+    if (this._inputMessagesEnabled) {
+      this._transport.disableInputMessages((err) => {
+        if (!err) {
+          this._inputMessagesEnabled = false;
+        }
+        callback(err);
+      });
+    } else {
+      callback();
+    }
+  }
+
+  private _enableInputMessages(callback: (err?: Error) => void): void {
+    if (!this._inputMessagesEnabled) {
+      const retryOp = new RetryOperation(this._retryPolicy, this._maxOperationTimeout);
+      retryOp.retry((opCallback) => {
+        /* Codes_SRS_NODE_MODULE_CLIENT_18_016: [ The client shall connect the transport if needed in order to receive inputMessages. ]*/
+        this._transport.enableInputMessages(opCallback);
+      }, (err) => {
+        if (!err) {
+          this._inputMessagesEnabled = true;
+        }
+        callback(err);
+      });
+    } else {
+      callback();
+    }
   }
 
-  /**
-   * @description       Creates an IoT Hub device client from the given
-   *                    connection string using the given transport type.
-   *
-   * @param {String}    connStr       A connection string which encapsulates "device
-   *                                  connect" permissions on an IoT hub.
-   * @param {Function}  Transport     A transport constructor.
-   *
-   * @throws {ReferenceError}         If the connStr parameter is falsy.
-   *
-   */
   static fromConnectionString(connStr: string, transportCtor: any): ModuleClient {
```

Comment:
>fromConnectionString [](start = 9, length = 20)

I don't necessarily like the way that Varun did his factory methods, but it feels like there should be a way to not copy/paste this code in both clients.  I don't know the right answer, but it's something to think about.

Reply:
there's only a bit of code that is shared. 2 things that motivated me to not refactor:
1. client constructors may evolve separately (including with separate parameters
2. not all auth mechanisms are supported everywhere which leads to minor differences

we could refactor argument checking (meh) and maybe instantiating the auth provider.. not that big of a gain imho. definitely a thing to keep in mind in the future though

---
In reply to: [195805036](https://github.com/Azure/azure-iot-sdk-node/pull/316#discussion_r195805036) [](ancestors = 195805036)